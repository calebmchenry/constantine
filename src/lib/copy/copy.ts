import { readFileSync, writeFileSync } from 'fs';
import { extname } from 'path';
import { getParser } from './parsers/parserFactory';
import { getComposer } from './composers/composerFactory';

function makeToFn(variableMap: any) {
    return (path: string): void => {
        // generate file
        const fileExtension = extname(path);
        const generateFn = getComposer(fileExtension);
        const fileContent = generateFn(variableMap);

        // write to file
        writeFileSync(path, fileContent);
    };
}

function makeWithFn(rawVariableMap: VariableMap) {
    return (transform: TransformFn) => {
        const transformedVariableMap = transform(rawVariableMap);
        return {
            to: makeToFn(transformedVariableMap),
        };
    };
}

function copy(path: string) {
    // read file
    const fileContent = readFileSync(path).toString();

    // generate variable map
    const fileExtension = extname(path);
    const parseFn = getParser(fileExtension);
    const variableMap = parseFn(fileContent);

    // make to function
    return {
        with: makeWithFn(variableMap),
        to: makeToFn(variableMap),
    };
}

export { copy };
