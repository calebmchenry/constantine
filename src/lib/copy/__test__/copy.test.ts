import fs from 'fs';
import { copy } from '../copy';

const mockedWriteFileSync = jest.fn();
fs.writeFileSync = mockedWriteFileSync;

test('copies js constants to js constants', () => {
    let toFileContent;
    mockedWriteFileSync.mockImplementation(
        (_path: string, fileContent: string) => {
            toFileContent = fileContent;
        }
    );
    copy('./src/lib/copy/__test__/examples/RestConstants.js').to(
        './some/path.js'
    );
    expect(toFileContent).toMatchSnapshot();
});

test('copies java constants to js constants', () => {
    let toFileContent;
    mockedWriteFileSync.mockImplementation(
        (_path: string, fileContent: string) => {
            toFileContent = fileContent;
        }
    );
    copy('./src/lib/copy/__test__/examples/RestConstants.java').to(
        './some/path.js'
    );
    expect(toFileContent).toMatchSnapshot();
});
