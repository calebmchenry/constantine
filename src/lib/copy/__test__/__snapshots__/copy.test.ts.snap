// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copies java constants to js constants 1`] = `
"export const FIZZ_PARAM = \\"fizz\\";
export const BAZ_PARAM = \\"baz\\";
export const FOO = \\"/foo\\";
export const FOO_BAR = \\"/foo/bar\\";
export const FOO_101_BAR = \\"/foo/101/bar\\";
export const FOO_FIZZ = FOO + \\"/{\\" + FIZZ_PARAM + \\"}\\";
export const POW = \\"/pow/\\"+BAZ_PARAM;
"
`;

exports[`copies js constants to js constants 1`] = `
"export const FIZZ_PARAM = 'fizz';
export const BAZ_PARAM = 'baz';
export const FOO = \\"/foo\\";
export const FOO_BAR = \`/foo/bar\`;
export const FOO_101_BAR = '/foo/101/bar';
export const FOO_FIZZ = FOO + \`/{\${ FIZZ_PARAM }}\`;
export const POW = \`/pow/\${{BAZ_PARAM}}\`;
export const FOO_FIZZ2 = FOO + '/{' + FIZZ_PARAM + '}';
export const POW2 = '/pow/{' + BAZ_PARAM  + \\"}\\";
"
`;
